File: src/main/java/com/google/android/finsky/api/model/PaginatedList.java
package com.google.android.finsky.api.model;

import android.text.TextUtils;

import com.android.volley.Request;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.google.android.finsky.protos.nano.Messages;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public abstract class PaginatedList<T, D> extends DfeModel implements Response.Listener<Messages.Response.ResponseWrapper>
{
    private final boolean autoLoadNextPage;
    private int currentOffset;
    private Request<?> currentRequest;
    private final List<D> items;
    private boolean itemsRemoved;
    private int itemsUntilEndCount;
    private int lastPositionRequested;
    Messages.Response.ResponseWrapper lastResponse;
    private boolean moreAvailable;
    List<UrlOffsetPair> urlOffsetList;
    private int windowDistance;
    
    PaginatedList(final String url) {
        this(url, true);
    }
    
    PaginatedList(final String url, final boolean autoLoadNextPage) {
        super();
        windowDistance = 12;
        items = new ArrayList<D>();
        itemsUntilEndCount = 4;
        (urlOffsetList = new ArrayList<>()).add(new UrlOffsetPair(0, url));
        moreAvailable = true;
        this.autoLoadNextPage = autoLoadNextPage;
    }
    
    PaginatedList(final List<UrlOffsetPair> urlOffsetList, final int count, final boolean autoLoadNextPage) {
        this(null, autoLoadNextPage);
        this.urlOffsetList = urlOffsetList;
        for (int i = 0; i < count; ++i) {
            this.items.add(null);
        }
    }
    
    private void requestMoreItemsIfNoRequestExists(final UrlOffsetPair urlOffsetPair) {
        if (!this.inErrorState()) {
            if (this.currentRequest != null && !this.currentRequest.isCanceled()) {
                if (this.currentRequest.getUrl().endsWith(urlOffsetPair.url)) {
                    return;
                }
                this.currentRequest.cancel();
            }
            this.currentOffset = urlOffsetPair.offset;
            this.currentRequest = this.makeRequest(urlOffsetPair.url);
        }
    }
    
    private void updateItemsUntilEndCount(final int n) {
        if (this.itemsUntilEndCount <= 0) {
            this.itemsUntilEndCount = 4;
            return;
        }
        this.itemsUntilEndCount = Math.max(1, n / 4);
    }
    
    public void clearDataAndReplaceInitialUrl(final String s) {
        this.urlOffsetList.clear();
        this.urlOffsetList.add(new UrlOffsetPair(0, s));
        this.resetItems();
    }
    
    protected void clearDiskCache() { }
    
    public void clearTransientState() {
        this.currentRequest = null;
    }
    
    public void flushUnusedPages() {
        if (this.lastPositionRequested >= 0) {
            for (int i = 0; i < this.items.size(); ++i) {
                if (i < -1 + (this.lastPositionRequested - this.windowDistance) || i >= this.lastPositionRequested + this.windowDistance) {
                    this.items.set(i, null);
                }
            }
        }
    }
    
    public int getCount() {
        return this.items.size();
    }
    
    public final D getItem(final int n) {
        return this.getItem(n, true);
    }
    
    public final D getItem(final int pos, final boolean isLastPosition) {
        if (isLastPosition) {
            this.lastPositionRequested = pos;
        }
        if (pos < 0) {
            throw new IllegalArgumentException("Can't return an item with a negative index: " + pos);
        }
        final int count = this.getCount();
        D value = null;
        if (pos < count) {
            value = this.items.get(pos);
            if (this.autoLoadNextPage && this.moreAvailable && pos >= this.getCount() - this.itemsUntilEndCount) {
                if (this.itemsRemoved) {
                    for (int i = 0; i < this.urlOffsetList.size(); ++i) {
                        if (this.urlOffsetList.get(i).offset > this.items.size()) {
                            while (this.urlOffsetList.size() > Math.max(1, i)) {
                                this.urlOffsetList.remove(-1 + this.urlOffsetList.size());
                            }
                            final UrlOffsetPair urlOffsetPair = this.urlOffsetList.get(-1 + this.urlOffsetList.size());
                            if (isLastPosition) {
                                this.requestMoreItemsIfNoRequestExists(urlOffsetPair);
                            }
                        }
                    }
                }
                else {
                    final UrlOffsetPair urlOffsetPair2 = this.urlOffsetList.get(-1 + this.urlOffsetList.size());
                    if (isLastPosition) {
                        this.requestMoreItemsIfNoRequestExists(urlOffsetPair2);
                    }
                }
            }
            if (value == null) {
                UrlOffsetPair urlOffsetPair3 = null;
                for (final UrlOffsetPair urlOffsetPair4 : this.urlOffsetList) {
                    if (urlOffsetPair4.offset > pos) {
                        break;
                    }
                    urlOffsetPair3 = urlOffsetPair4;
                }
                this.requestMoreItemsIfNoRequestExists(urlOffsetPair3);
            }
        }
        return value;
    }
    
    protected abstract D[] getItemsFromResponse(final Messages.Response.ResponseWrapper listResponse);
    
    public List<String> getListPageUrls() {
        final ArrayList<String> list = new ArrayList<String>(this.urlOffsetList.size());
        final Iterator<UrlOffsetPair> iterator = this.urlOffsetList.iterator();
        while (iterator.hasNext()) {
            list.add(iterator.next().url);
        }
        return list;
    }
    
    protected abstract String getNextPageUrl(final Messages.Response.ResponseWrapper listResponse);
    
    public boolean isMoreAvailable() {
        return this.moreAvailable;
    }
    
    @Override
    public boolean isReady() {
        return this.lastResponse != null || this.items.size() > 0;
    }
    
    protected abstract Request<?> makeRequest(final String url);
    
    @Override
    public void onErrorResponse(final VolleyError volleyError) {
        this.clearTransientState();
        super.onErrorResponse(volleyError);
    }

    @Override
    public void onResponse(final Messages.Response.ResponseWrapper lastResponse) {
        this.clearErrors();
        this.lastResponse = lastResponse;
        final int size = this.items.size();
        final D[] itemsFromResponse = this.getItemsFromResponse(lastResponse);
        this.updateItemsUntilEndCount(itemsFromResponse.length);
        for (int i = 0; i < itemsFromResponse.length; ++i) {
            if (i + this.currentOffset < this.items.size()) {
                this.items.set(i + this.currentOffset, (D)itemsFromResponse[i]);
            }
            else {
                this.items.add(itemsFromResponse[i]);
            }
        }
        final String nextPageUrl = this.getNextPageUrl(lastResponse);
        if (!TextUtils.isEmpty(nextPageUrl) && (this.currentOffset == size || this.itemsRemoved)) {
            this.urlOffsetList.add(new UrlOffsetPair(this.items.size(), nextPageUrl));
        }
        if (this.itemsRemoved) {
            this.itemsRemoved = false;
        }
        final int offset = this.urlOffsetList.get(-1 + this.urlOffsetList.size()).offset;
        boolean moreAvailable = false;
        if (items.size() == offset) {
            moreAvailable = (itemsFromResponse.length > 0);
        }
        this.moreAvailable = (moreAvailable && autoLoadNextPage);
        this.clearTransientState();
        this.notifyDataSetChanged();
    }
    
    public void removeItem(final int n) {
        this.items.remove(n);
        this.itemsRemoved = true;
        if (this.currentRequest != null && !this.currentRequest.isCanceled()) {
            this.currentRequest.cancel();
        }
        this.clearDiskCache();
    }
    
    public void resetItems() {
        this.moreAvailable = true;
        this.items.clear();
        this.notifyDataSetChanged();
    }
    
    public void retryLoadItems() {
        if (this.inErrorState()) {
            this.clearTransientState();
            this.clearErrors();
            final int mCurrentOffset = this.currentOffset;
            UrlOffsetPair urlOffsetPair = null;
            Label_0078: {
                if (mCurrentOffset != -1) {
                    final Iterator<UrlOffsetPair> iterator = this.urlOffsetList.iterator();
                    UrlOffsetPair urlOffsetPair2;
                    do {
                        final boolean hasNext = iterator.hasNext();
                        urlOffsetPair = null;
                        if (!hasNext) {
                            break Label_0078;
                        }
                        urlOffsetPair2 = iterator.next();
                    } while (this.currentOffset != urlOffsetPair2.offset);
                    urlOffsetPair = urlOffsetPair2;
                }
            }
            if (urlOffsetPair == null) {
                urlOffsetPair = this.urlOffsetList.get(-1 + this.urlOffsetList.size());
            }
            this.requestMoreItemsIfNoRequestExists(urlOffsetPair);
        }
    }

    public void startLoadItems() {
        if (this.moreAvailable && this.getCount() == 0) {
            this.clearErrors();
            this.requestMoreItemsIfNoRequestExists(this.urlOffsetList.get(0));
        }
    }
    
    protected static class UrlOffsetPair
    {
        public final int offset;
        public final String url;
        
        public UrlOffsetPair(final int offset, final String url) {
            super();
            this.offset = offset;
            this.url = url;
        }
    }
}

File: playstore/src/main/java/finsky/api/model/PaginatedList.kt
package finsky.api.model

import com.android.volley.Request
import com.android.volley.Response
import com.android.volley.VolleyError
import finsky.protos.Messages
import kotlinx.coroutines.channels.Channel
import java.util.*
import kotlin.math.max

class UrlOffsetPair(val offset: Int, val url: String)

class ListAvailable(val isFirst: Boolean, val requestedMoreItems: Boolean, val error: VolleyError?)

typealias FilteredDocumentList<D> = Pair<Array<D>, Int>

abstract class PaginatedList<T, D>(
        private val urlOffsetList: MutableList<UrlOffsetPair>,
        private val autoLoadNextPage: Boolean = true) : DfeModel(), Response.ErrorListener, Response.Listener<Messages.Response.ResponseWrapper> {
    private var currentOffset: Int = 0
    private var currentRequest: Request<*>? = null
    private val items = mutableListOf<D>()
    private var itemsRemoved: Boolean = false
    private var itemsUntilEndCount: Int = 4
    private var lastPositionRequested: Int = 0
    private var lastResponse: Messages.Response.ResponseWrapper? = null
    private var isMoreAvailable: Boolean = true
    private val windowDistance: Int = 12
    private var isInErrorState = false

    val updates = Channel<ListAvailable>(Channel.CONFLATED)

    constructor(url: String, autoLoadNextPage: Boolean)
            : this(mutableListOf(UrlOffsetPair(0, url)), autoLoadNextPage)

    private fun requestMoreItemsIfNoRequestExists(urlOffsetPair: UrlOffsetPair?) {
        if (!this.isInErrorState) {
            if (this.currentRequest != null && !this.currentRequest!!.isCanceled) {
                if (this.currentRequest!!.url.endsWith(urlOffsetPair!!.url)) {
                    return
                }
                this.currentRequest!!.cancel()
            }
            this.currentOffset = urlOffsetPair!!.offset
            this.currentRequest = this.makeRequest(urlOffsetPair.url, this, this)
        }
    }

    private fun updateItemsUntilEndCount(n: Int) {
        if (this.itemsUntilEndCount <= 0) {
            this.itemsUntilEndCount = 4
            return
        }
        this.itemsUntilEndCount = max(1, n / 4)
    }

    private fun clearTransientState() {
        this.currentRequest = null
    }

    val count: Int
        get() = this.items.size

    fun getItem(pos: Int, isLastPosition: Boolean): D? {
        if (isLastPosition) {
            this.lastPositionRequested = pos
        }
        require(pos >= 0) { "Can't return an item with a negative index: $pos" }
        val count = this.count
        var value: D? = null
        if (pos < count) {
            value = this.items[pos]
            if (this.autoLoadNextPage && this.isMoreAvailable && pos >= this.count - this.itemsUntilEndCount) {
                if (this.itemsRemoved) {
                    for (i in this.urlOffsetList.indices) {
                        if (this.urlOffsetList[i].offset > this.items.size) {
                            while (this.urlOffsetList.size > max(1, i)) {
                                this.urlOffsetList.removeAt(-1 + this.urlOffsetList.size)
                            }
                            val urlOffsetPair = this.urlOffsetList[-1 + this.urlOffsetList.size]
                            if (isLastPosition) {
                                this.requestMoreItemsIfNoRequestExists(urlOffsetPair)
                            }
                        }
                    }
                } else {
                    val urlOffsetPair2 = this.urlOffsetList[-1 + this.urlOffsetList.size]
                    if (isLastPosition) {
                        this.requestMoreItemsIfNoRequestExists(urlOffsetPair2)
                    }
                }
            }
            if (value == null) {
                var urlOffsetPair3: UrlOffsetPair? = null
                for (urlOffsetPair4 in this.urlOffsetList) {
                    if (urlOffsetPair4.offset > pos) {
                        break
                    }
                    urlOffsetPair3 = urlOffsetPair4
                }
                this.requestMoreItemsIfNoRequestExists(urlOffsetPair3)
            }
        }
        return value
    }

    val listPageUrls: List<String>
        get() {
            val list = ArrayList<String>(this.urlOffsetList.size)
            val iterator = this.urlOffsetList.iterator()
            while (iterator.hasNext()) {
                list.add(iterator.next().url)
            }
            return list
        }

    protected abstract fun getNextPageUrl(wrapper: Messages.Response.ResponseWrapper): String?
    protected abstract fun getItemsFromResponse(wrapper: Messages.Response.ResponseWrapper): FilteredDocumentList<D>

    override val isReady: Boolean
        get() = this.lastResponse != null || this.items.size > 0

    override fun onErrorResponse(error: VolleyError) {
        this.isInErrorState = true
        this.clearTransientState()
        val isFirst = this.urlOffsetList.size == 1
        updates.offer(ListAvailable(isFirst, false, error))
    }

    override fun onResponse(responseWrapper: Messages.Response.ResponseWrapper) {
        this.lastResponse = responseWrapper
        val size = this.items.size
        val itemsFromResponse = this.getItemsFromResponse(responseWrapper)
        this.updateItemsUntilEndCount(itemsFromResponse.first.size)
        val isFirst = this.urlOffsetList.size == 1
        for (i in itemsFromResponse.first.indices) {
            if (i + this.currentOffset < this.items.size) {
                this.items[i + this.currentOffset] = itemsFromResponse.first[i]
            } else {
                this.items.add(itemsFromResponse.first[i])
            }
        }
        val nextPageUrl = this.getNextPageUrl(responseWrapper)
        var urlOffsetPair2: UrlOffsetPair? = null
        if (!nextPageUrl.isNullOrEmpty() && (this.currentOffset == size || this.itemsRemoved)) {
            urlOffsetPair2 = UrlOffsetPair(this.items.size, nextPageUrl)
            this.urlOffsetList.add(urlOffsetPair2)
        }
        if (this.itemsRemoved) {
            this.itemsRemoved = false
        }
        val offset = this.urlOffsetList[-1 + this.urlOffsetList.size].offset
        var moreAvailable = false
        if (items.size == offset) {
            moreAvailable = itemsFromResponse.second > 0
        }
        this.isMoreAvailable = moreAvailable && autoLoadNextPage
        this.clearTransientState()
        val requestMoreItems = moreAvailable && urlOffsetPair2 != null
        if (requestMoreItems) {
            this.requestMoreItemsIfNoRequestExists(urlOffsetPair2)
        }
        updates.offer(ListAvailable(isFirst, requestMoreItems, null))
    }

    fun resetItems() {
        this.isMoreAvailable = true
        this.items.clear()
        val isFirst = this.urlOffsetList.size == 1
        updates.offer(ListAvailable(isFirst, false, null))
    }

    fun startLoadItems() {
        if (this.isMoreAvailable && this.count == 0) {
            this.isInErrorState = false
            this.requestMoreItemsIfNoRequestExists(this.urlOffsetList[0])
        }
    }
}
